### PR TITLE
BACK-742: Support Fee Tokens

### PR DESCRIPTION
**Pull request by Verisana**

### CODE REVIEW

File: src/dex/curve-v1/curve-v1.ts [hunk 1]
```diff
@@ -681,41 +695,46 @@ export class CurveV1 extends SimpleExchange implements IDex<CurveV1Data> {
...
-      if (side === SwapSide.BUY) {
+      const isSell = side === SwapSide.SELL;
+      if (!isSell) {
         return null;
       }
-      const from = this.dexHelper.config.wrapETH(_from);
-      const to = this.dexHelper.config.wrapETH(_to);
+
+      const _isTokenTransferFeeToBeExchanged =
+        isTokenTransferFeeToBeExchanged(transferFees);
```
Comment (ColonelJ) [inline src/dex/curve-v1/curve-v1.ts:712]: I think we don't care about whether destToken has fees, only srcToken.

File: src/dex/curve-v1/curve-v1.ts [hunk 2]
```diff
@@ -681,41 +695,46 @@ export class CurveV1 extends SimpleExchange implements IDex<CurveV1Data> {
...
+        } else {
+          const id = this.getPoolIdentifier(p.name);
+          keepPool = limitPools!.includes(id);
+        }
+
+        if (_isTokenTransferFeeToBeExchanged) {
+          // Fee on transfers supported only when flag is specified
+          return keepPool && !!p.isFeeOnTransferSupported;
+        } else {
+          return keepPool;
```
Comment (ColonelJ) [inline src/dex/curve-v1/curve-v1.ts:730]: It doesn't make sense to have this `keepPool` variable, when you can simply `return false;` at the appropriate places and `return true;` at the end.  Please refactor.

File: src/dex/uniswap-v2/uniswap-v2.ts [hunk 3]
```diff
@@ -537,32 +540,56 @@ export class UniswapV2
...
+
+      const outputsWithFee = applyTransferFee(
+        applyTransferFee(
+          prices,
+          side,
+          isSell ? transferFees.destFee : transferFees.srcFee,
+          isSell ? DEST_TOKEN_PARASWAP_TRANSFERS : SRC_TOKEN_PARASWAP_TRANSFERS,
+        ),
+        side,
+        isSell ? transferFees.destDexFee : transferFees.srcDexFee,
```
Comment (ColonelJ) [inline src/dex/uniswap-v2/uniswap-v2.ts:586]: `srcDexFee` should not be applied as it is included in the pool fee accounted for in pricing.

File: src/dex/uniswap-v2/uniswap-v2.ts [hunk 4]
```diff
@@ -537,32 +540,56 @@ export class UniswapV2
...
-      const unitAmount = getBigIntPow(
-        side == SwapSide.BUY ? to.decimals : from.decimals,
+      const unitAmount = getBigIntPow(isSell ? from.decimals : to.decimals);
+
+      const [unitWithFee, ...amountsWithFee] = applyTransferFee(
+        [unitAmount, ...amounts],
+        side,
+        transferFees.srcFee,
+        SRC_TOKEN_PARASWAP_TRANSFERS,
       );
```
Comment (ColonelJ) [inline src/dex/uniswap-v2/uniswap-v2.ts:560]: Need to account for BUY case, where buyFee and buyDexFee should be applied.

File: src/pricing-helper.ts [hunk 5]
```diff
@@ -143,6 +144,26 @@ export class PricingHelper {
     );
   }
 
+  getPoolsSupportingFeeOnTransfer(): string[] {
```
Comment (ColonelJ) [inline src/pricing-helper.ts:147]: Badly named, should be `getDexs`?

**Review by ColonelJ — CHANGES_REQUESTED** [review]

File: src/pricing-helper.ts (hunk 5 above)
Comment (Verisana) [reply src/pricing-helper.ts]: Yeah, fixed

File: src/dex/uniswap-v2/uniswap-v2.ts (hunk 4 above)
Comment (Verisana) [reply src/dex/uniswap-v2/uniswap-v2.ts]: I think I fixed it

File: src/dex/uniswap-v2/uniswap-v2.ts (hunk 3 above)
Comment (Verisana) [reply src/dex/uniswap-v2/uniswap-v2.ts]: You are right, replaced with 0

File: src/dex/curve-v1/curve-v1.ts (hunk 1 above)
Comment (Verisana) [reply src/dex/curve-v1/curve-v1.ts]: Yeah, fixed this as well

**Review by Louis-Amas — APPROVED** [review]

**Review by ColonelJ — APPROVED** [review]

**Pull request merged**